### PR TITLE
hare: unstable-2023-04-23 -> unstable-2023-10-22; harec: unstable-2023-04-23 -> unstable-2023-10-23

### PR DESCRIPTION
**Pull request by onemoresuza**
## Description of changes

Update hare and harec to the refs on Alpine and do the following refactoring:

### hare:

1. Replace `NIX_BUILD_TOP` usage with `mktemp`
2. Enable parallel building
3. Get rid of `config-template.mk`, by using `makeFlagsArray` instead
4. Move `wrapProgram` call to `postFixup`
5. Set the `LOCALVER` environment variable to `nix`, so that the `hare
   version` command outputs `dev-nix`[1]
6. Set `meta` attribute `mainProgram`
7. Make the package accessible through `all-packages.nix` instead of
   `aliases.nix`
8. Move package path from `pkgs/development/compilers/hare/hare` to
   `pkgs/development/compilers/hare`, since the scope is now made at
   `pkgs/top-level/hare-packages.nix`

### harec:

1. Remove `hardeningDisable = [ "fortify" ];`, since both harec and hare
   compile normally with it enabled
2. Enable parallel building
3. Set `meta` attribute `mainProgram`
4. Make the package accessible through `all-packages.nix` instead of
   `aliases.nix`
5. Move package path from `pkgs/development/compilers/hare/harec` to
   `pkgs/development/compilers/harec`, since the scope is now made at
    `pkgs/top-level/hare-packages.nix`

### harePackages:

1. Move hare packages scope from
   `pkgs/development/compilers/hare/default.nix` to
   `pkgs/top-level/hare-packages.nix`
2. Remove `hare` and `harec` from `aliases.nix`, moving them to
   `all-packages.nix`
3. Change the `callPackage` argument in `all-packages.nix` from
   `../development/compilers/hare` to `./hare-packages.nix`

[1]: https://git.sr.ht/~sircmpwn/hare/tree/1048620a7a25134db370bf24736efff1ffcb2483/item/scripts/version#L2

Co-authored-by: starzation <nixpkgs@starzation.net>

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by onemoresuza** [comment]
Output of `nixpkgs-review rev HEAD`:

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>himitsu</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>harePackages.hare</li>
    <li>harePackages.harec</li>
  </ul>
</details>

The `himitsu` package fails to build because of some changes to Hare's `stdlib`;
when updated to [0.4](https://git.sr.ht/~sircmpwn/himitsu/refs/0.4), It will
work again.

**Review by majiru — APPROVED** [review]
The comments around the temporary changes are nice and the refactoring seems like an improvement to me.

**Comment by pbsds** [comment]
Due to the himitsu breakage this is likely blocked until after the 23.11 branch-of, since himitsu 0.4 looks to be a breaking change. Feel free to ping me after the branch-of :smiley:

**Comment by onemoresuza** [comment]
> Due to the himitsu breakage this is likely blocked until after the 23.11 branch-of, since himitsu 0.4 looks to be a breaking change. Feel free to ping me after the branch-of 😃

No worries. Should I do something to this PR to indicate that it is waiting for the `23.11` release? Perhaps renaming the title or adding a label?

**Comment by pbsds** [comment]
I don't think we have a convention for such. I've been toying with the idea of proposing a post-ZHF merge queue, and a wait-for-maintainer-feedback-but-a-committer-wants-to-merge-it merge queue.

**Comment by onemoresuza** [comment]
@majiru, @pbsds, I was thinking about the changes I've made to `qbe` on both `hare` and `harec`. 

As of now:

1. if the user overrides the `qbe` argument of `harec` with `override`, any change made to its `src` or `version` would not matter, since they are locally overwritten.
1. Moreover, the change made to `hare` will cause breakage to any code that needs the `qbe` argument on it, since it is now removed.

Are those acceptable changes?

**Comment by starzation** [comment]
@onemoresuza Maybe you should update this to latest version

**Comment by onemoresuza** [comment]
No worries, @starzation.

I'm planning to do this, but firstly this [PR](https://github.com/NixOS/nixpkgs/pull/265124#issue-1975311634) must be merged: the most recent versions of `Hare` need the `leap-seconds.list` file for both its test suite and its [timekeeping needs](https://harelang.org/distributions/).

**Review by starzation — CHANGES_REQUESTED** [review]
`homepage` should be use `https`, other than these is good

File: pkgs/development/compilers/hare/harec/default.nix
Comment (starzation) [reply pkgs/development/compilers/hare/harec/default.nix:51]: I forgot there are another in `hare` should change that too

Comment (onemoresuza) [reply pkgs/development/compilers/hare/harec/default.nix]: Fixed

**Review by starzation — APPROVED** [review]

**Comment by pbsds** [comment]
> Are those acceptable changes?

I think this is fine. It's still overrideable, it just requires a bit more effort. And there is always the possibility to change this later

@ofborg build hare hare.passthru.tests harec harec.passhtru.tests

**Comment by onemoresuza** [comment]
@starzation, I think this PR would be a good place for the creation of the `pkgs/top-level/hare-packages.nix` scope that you created in your [PR for hare-json](https://github.com/NixOS/nixpkgs/pull/268982#issue-2004524346).

Would you be okay if I brought that change to this PR and add you as a co-author of this commit? If so, could you please give the email to be appended to your username on the git trailer, *i. e.*, either a real email or the github no-reply one[^1]?

[^1]: Which can be found here <https://github.com/settings/emails> under the `Keep my email addresses private` checkbox.

**Comment by starzation** [comment]
> @starzation, I think this PR would be a good place for the creation of the `pkgs/top-level/hare-packages.nix` scope that you created in your [PR for hare-json](https://github.com/NixOS/nixpkgs/pull/268982#issue-2004524346).
> 
> Would you be okay if I brought that change to this PR and add you as a co-author of this commit? If so, could you please give the email to be appended to your username on the git trailer, _i. e._, either a real email or the github no-reply one[1](#user-content-fn-1-d983b488c108cc4973a256603d5f9d47)?
> ## Footnotes
> 
>     1. Which can be found here https://github.com/settings/emails under the `Keep my email addresses private` checkbox. [↩](#user-content-fnref-1-d983b488c108cc4973a256603d5f9d47)

Sure, I'm okay with that. You can use this email address `nixpkgs@starzation.net`

File: pkgs/development/compilers/hare/hare/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/hare/hare/default.nix:7]:
```suggestion
```
comments in inputs are often overlooked

File: pkgs/development/compilers/hare/hare/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/hare/hare/default.nix:13]:
```suggestion
  # we use harec's override of qbe until 1.2 is released
  qbe = harec.qbeUnstable;
```

File: pkgs/development/compilers/hare/hare/default.nix
Comment (SuperSandro2000) [inline pkgs/development/compilers/hare/hare/default.nix:69]: Those don't work in makeFlags?

**Review by SuperSandro2000 — COMMENTED** [review]
otherwise LGTM

File: pkgs/development/compilers/hare/hare/default.nix
Comment (onemoresuza) [reply pkgs/development/compilers/hare/hare/default.nix:69]: With the exception of `HARECACHE` and `BINOUT`, they would. I've put them all together in `makeFlagsArray` so that the make flags are not split between `makeFlags` and `makeFlagsArray`.

But if it the ones that work on `makeFlags` should necessarily be in it, I can move them there.

File: pkgs/development/compilers/hare/hare/default.nix
Comment (onemoresuza) [reply pkgs/development/compilers/hare/hare/default.nix]: I've moved the comment to the `qbe` variable in the `let-in` statement, while also keeping the `qbe` argument to avoid breakage with code that passes it to `hare`.

File: pkgs/development/compilers/hare/hare/default.nix
Comment (onemoresuza) [reply pkgs/development/compilers/hare/hare/default.nix]: Done --- do note that the file was changed from `pkgs/development/compilers/hare/hare/default.nix` to `pkgs/development/compilers/hare/default.nix`, as per the explanation in the updated description of this PR

**Comment by onemoresuza** [comment]
The `ofborg` error was on the [`ofborg-eval-nixpkgs-manual`](https://gist.github.com/GrahamcOfBorg/bd13769eee5f5c82a074e30742fb45e9). I don't think it's related to the pushed changes since the error occurs `while evaluating the attribute 'buildPhase' of the derivation 'nixpkgs-manual'`, which I believe is not affected by this PR.

Any thoughts on this?

File: pkgs/development/compilers/hare/hare/default.nix
Comment (onemoresuza) [reply pkgs/development/compilers/hare/hare/default.nix]: A note: github's `outdated` warning is because the file path has changed from `pkgs/development/compilers/hare/hare/default.nix` to `pkgs/development/compilers/hare/default.nix`. The `makeFlagsArray` has not changed.

**Comment by starzation** [comment]
@ofborg eval

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 266456` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>himitsu-firefox</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>himitsu</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>hare (harePackages.hare)</li>
    <li>harec (harePackages.harec)</li>
  </ul>
</details>

Himitsu to be fixed in #267060, LGTM, should not be backported. Thanks!

**Pull request merged**
